File: lib/day11.ex
require Integer

defmodule Day11 do
  def count(stone, steps, cache \\ %{}) do
    case Map.get(cache, {stone, steps}) do
      nil ->
        result = _count(stone, steps, cache)
        Map.put(cache, {stone, steps}, result)

        result

      result ->
        result
    end
  end

  def load_input(filename) do
    File.read!(filename)
    |> String.split("\s", trim: true)
    |> Enum.map(&String.to_integer/1)
  end

  def part1(stones) do
    Enum.map(stones, fn stone ->
      count(stone, 25)
    end)
    |> Enum.sum()
  end

  def part2(stones) do
    Enum.map(stones, fn stone ->
      count(stone, 75)
    end)
    |> Enum.sum()
  end

  defp _count(_stone, 0, _cache), do: 1
  defp _count(0, steps, cache), do: count(1, steps - 1, cache)

  defp _count(stone, steps, cache) do
    if even_number_of_digits?(stone) do
      split_in_half(stone)
      |> Enum.map(fn n -> count(n, steps - 1, cache) end)
      |> Enum.sum()
    else
      count(stone * 2024, steps - 1, cache)
    end
  end

  # def blink(stones, steps) do
  #   Enum.flat_map(stones, fn stone ->
  #     blink_stone(stone, steps)
  #   end)
  # end

  # defp blink_stone(stone, steps) do
  #   if even_number_of_digits?(stone) do
  #     split_in_half(stone)
  #   else
  #     [n * 2024]
  #   end
  # end

  defp even_number_of_digits?(n) do
    Integer.digits(n)
    |> length()
    |> Integer.is_even()
  end

  defp split_in_half(n) do
    digits = Integer.digits(n)
    half = div(length(digits), 2)
    {left, right} = Enum.split(digits, half)

    [left, right]
    |> Enum.map(fn digits ->
      digits |> Enum.join() |> String.to_integer()
    end)
  end
end

File: day11.rb
require "benchmark" 
class Day11
  attr_accessor :stones, :counts

    def self.part1(stones)
     new(stones, 25).blink
    end
    def self.part2(stones)
     new(stones, 75).blink
    end

  def initialize(stones, steps)
    @stones = stones
    @steps = steps
    @cache = {}
    stones.each do |stone|
      @cache[stone.to_s] = {"#{steps}": 1 }
    end
  end

  def blink
    @stones.map { |stone| count(stone, @steps) }.sum
  end

  def count(stone,  steps)
    if @cache.fetch(stone.to_s, {}).fetch(steps.to_s, nil)
      print "c"
      return @cache[stone.to_s][steps.to_s]
    else
    print "."
    end

    if steps.zero?
      result = 1
    elsif stone.zero?
      result = count(1, steps - 1)

    elsif stone.to_s.size.even?
      half = stone.to_s.size / 2
      left, right = stone.to_s.chars.each_slice(half).map(&:join).map(&:to_i)
      result = count(left, steps - 1) + count(right, steps - 1)
    else
      result = count(stone * 2024, steps - 1)
    end
    cache(stone, steps, result)
    result 
  end

  def cache(stone, steps, result)
    @cache[stone.to_s] ||= {}
    @cache[stone.to_s][steps.to_s] = result
  end
end

input = [8793800,1629,65,5,960,0,138983,85629]
puts Benchmark.measure { puts Day11.new(input, 7250).blink }


